initrd/kraftignore.go

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2022, Unikraft GmbH and The KraftKit Authors.
+// Licensed under the BSD-3-Clause License (the "License").
+// You may not use this file except in compliance with the License.
+package initrd
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"kraftkit.sh/log"
+)
+
+// kraftignore filename
+const KraftignoreFileName = ".kraftignore"
+
+type IgnoringFileType string
+
+const (
+	Exist    = IgnoringFileType("Exist")
+	NotExist = IgnoringFileType("NotExist")
+	SkipDir  = IgnoringFileType("SkipDir")
+)
+
+func getKraftignoreItems(ctx context.Context, initrd *directory) ([]string, error) {
+	if initrd.opts.kraftignorePath == "" {
+		cwd, err := os.Getwd()
+		if err != nil {
+			return []string{}, err
+		}
+		initrd.opts.kraftignorePath = filepath.Join(cwd, KraftignoreFileName)
+	}
+
+	if _, err := os.Stat(initrd.opts.kraftignorePath); errors.Is(err, os.ErrNotExist) {
+		return []string{}, nil
+	} else if err != nil {
+		return []string{}, err
+	}
+
+	kraftignoreFile, err := os.Open(initrd.opts.kraftignorePath)
+	if err != nil {
+		return []string{}, err
+	}
+
+	defer func() {
+		kraftIgnoreErr := kraftignoreFile.Close()
+		if kraftIgnoreErr != nil {
+			if err != nil {
+				err = fmt.Errorf("%v\n%v", err, kraftIgnoreErr)
+			} else {
+				err = kraftIgnoreErr
+			}
+		}
+	}()
+
+	kraftignoreScanner := bufio.NewScanner(kraftignoreFile)
+	kraftignoreScanner.Split(bufio.ScanLines)
+	var kraftignoreFileLines, ignoringItems []string
+
+	for kraftignoreScanner.Scan() {
+		kraftignoreFileLines = append(kraftignoreFileLines, kraftignoreScanner.Text())
+	}
+
+	for lineNum, line := range kraftignoreFileLines {
+		line = strings.TrimSpace(line)
+		if line == "" || strings.HasPrefix(line, "#") {
+			continue
+		}
+		items := strings.Split(line, " ")
+		for _, item := range items {
+			item = strings.Trim(item, `"`)
+			item = strings.Trim(item, `'`)
+			item = strings.TrimSpace(item)
+			if item == "" || item == "#" {
+				continue
+			}
+
+			if hasGlobPatterns(item) {
+				log.G(ctx).
+					WithField("file", initrd.opts.kraftignorePath).
+					Warn("contains a glob pattern ", item,
+						" at line ", lineNum,
+						" which is not supported by Kraftkit")
+				continue
+			}
+
+			item = strings.TrimPrefix(item, "..")
```

Comment:
Same here, I don't think we should trim this, but rather what I would do would be:
1. Find the full path of the initrd directory
1. Append the file path to the initrd directory
1. Check if it evaluates to a valid file, if yes, trim `1.` from it and use that given path, if not, ignore

This goes for the whole file parsing logic

Reply:
Updated the logic to work in your recommended way, Please review again.

But still removing `..` if item contains `../` as prefix because path inside `rootfs` which starts `../xyz` refer to the content outside of `rootfs` dir.

I hope this make sense otherwise ping me again ;)